### PR TITLE
s4696681: Implementation of GCN Model for Node Classification on the page-page Facebook dataset

### PR DESCRIPTION
**Pull request by kaia-santosha**
This is my pull request to merge the node classification facebook page-page problem for the Report Assessment for COMP3710

My implementation is a Graph Convolutional Network for the purpose of classifying nodes in the dataset

My implementation does the following:
* imports and preprocesses the facebook page-page dataset
   - Creates an adjacency matrix from the data provided in musae_facebook_edges.csv
   - Normalises this adjacency matrix as required by the GCN
   - creates the feature vectors for each node, since there were an inconsistent number of features per node, bag of word feature vectors were created
   - encodes classifications to be integers instead of strings so they can be used in the model
   - create the tensors required by pytorch to be fed into the model

The model itself consists of two Graph Convolutional Layers and the model is tuned and trained using nested cross validation

This pull request contains the required files to satisfy the Task Sheet:
modules.py (contains model architecture)
dataset.py (contains preprocessing of data)
train.py (contains the training for the model)
predict.py (is the test driver, where through this all the other files are called)
README.md (contains information about the project and how to run the model)

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Normal

Readme Overall: Good

- Project Overview: Good
- Model: Good
- Data + Preprocessing: Good
- Training / Loss Curve(s): **None**
- Result Demonstration: Good
- References: Good

Functionality:

- Multi-layer GCN: Good, 2 layers.
- Semi supervised: Good
- Accuracy: OK. Not mentioned if its Test or Train accuracy.
- TSNE or UMAP: Good, though would have expected to see a bit more separation at them mentioned (91%) accuracy.
- Discussion: Good

Code: Has early stop in the code, but also uses multiple epochs as a searchable hyper parameter, which is redundant.

- Consistent with Results: OK.
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
Epochs should not need to be in the hyper parameter search. You do have an early stop in the code anyways. You could have used the number of GCN layers as a hyper parameter.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage, no training plots -2
Module present
Commenting minimal -1
No Data leakage
Difficulty: Nromal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable, no refs -1, no usage -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

**Comment by shakes76** [comment]
Restore repo README to allow merge, does not affect grade only merge.

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks lost.
